File: packages/sdk/hush-ts/src/policy-lab.ts
import { getWasmModule } from "./crypto/backend.js";

export interface SynthResult {
  policyYaml: string;
  risks: string[];
}

export interface SimulationSummary {
  total: number;
  allowed: number;
  warn: number;
  blocked: number;
}

export interface SimulationResultEntry {
  eventId: string;
  outcome: string;
  allowed: boolean;
  denied: boolean;
  warn: boolean;
  reasonCode: string;
  guard?: string;
  severity?: string;
  message?: string;
}

export interface SimulateResult {
  summary: SimulationSummary;
  results: SimulationResultEntry[];
}

/**
 * PolicyLab: unified observe -> hunt -> OCSF -> synth pipeline.
 * Backed by Rust compiled to WASM.
 * Requires `initWasm()` before construction.
 */
export class PolicyLab {
  // biome-ignore lint/suspicious/noExplicitAny: WasmPolicyLab is untyped
  private readonly inner: any;

  constructor(policyYaml: string) {
    const wasm = getWasmModule();
    if (!wasm?.WasmPolicyLab) {
      throw new Error(
        "WASM not initialized. Call initWasm() before using PolicyLab.",
      );
    }
    this.inner = new wasm.WasmPolicyLab(policyYaml);
  }

  /** Simulate events against the loaded policy. */
  simulate(eventsJsonl: string): SimulateResult {
    const json: string = this.inner.simulate(eventsJsonl);
    return JSON.parse(json) as SimulateResult;
  }

  /** Synthesize a candidate policy from observed events. */
  static synth(eventsJsonl: string): SynthResult {
    const wasm = getWasmModule();
    if (!wasm?.policy_lab_synth) {
      throw new Error(
        "WASM not initialized. Call initWasm() before using PolicyLab.",
      );
    }
    const json: string = wasm.policy_lab_synth(eventsJsonl);
    return JSON.parse(json) as SynthResult;
  }

  /** Convert PolicyEvent JSONL to OCSF JSONL. */
  static toOcsf(eventsJsonl: string): string {
    const wasm = getWasmModule();
    if (!wasm?.policy_lab_to_ocsf) {
      throw new Error(
        "WASM not initialized. Call initWasm() before using PolicyLab.",
      );
    }
    return wasm.policy_lab_to_ocsf(eventsJsonl);
  }

  /** Convert PolicyEvent JSONL to TimelineEvent JSONL. */
  static toTimeline(eventsJsonl: string): string {
    const wasm = getWasmModule();
    if (!wasm?.policy_lab_to_timeline) {
      throw new Error(
        "WASM not initialized. Call initWasm() before using PolicyLab.",
      );
    }
    return wasm.policy_lab_to_timeline(eventsJsonl);
  }
}

File: packages/sdk/hush-py/src/clawdstrike/policy_lab.py
"""PolicyLab: unified observe -> hunt -> OCSF -> synth pipeline (native-backed).

Requires the hush_native extension with PolicyLab support.
"""

from __future__ import annotations

from typing import Any

from clawdstrike.native import NATIVE_AVAILABLE, get_native_module


def _require_native() -> None:
    """Fail-closed: ensure native PolicyLab bindings are available."""
    if not NATIVE_AVAILABLE:
        raise ImportError(
            "clawdstrike.policy_lab requires the native extension (hush-native). "
            "Build/install from `packages/sdk/hush-py/hush-native`."
        )
    mod = get_native_module()
    if not hasattr(mod, "PolicyLab"):
        raise ImportError(
            "hush-native does not include PolicyLab bindings. "
            "Rebuild with latest hush-native."
        )


class PolicyLab:
    """Unified observe -> hunt -> OCSF -> synth pipeline."""

    def __init__(self, policy_yaml: str) -> None:
        _require_native()
        mod = get_native_module()
        self._inner = mod.PolicyLab(policy_yaml)

    def simulate(self, events_jsonl: str) -> dict[str, Any]:
        """Simulate events against the loaded policy."""
        return self._inner.simulate(events_jsonl)

    @staticmethod
    def synth(events_jsonl: str) -> dict[str, Any]:
        """Synthesize a candidate policy from observed events."""
        _require_native()
        mod = get_native_module()
        return mod.PolicyLab.synth(events_jsonl)

    @staticmethod
    def to_ocsf(events_jsonl: str) -> str:
        """Convert PolicyEvent JSONL to OCSF JSONL."""
        _require_native()
        mod = get_native_module()
        return mod.PolicyLab.to_ocsf(events_jsonl)

    @staticmethod
    def to_timeline(events_jsonl: str) -> str:
        """Convert PolicyEvent JSONL to TimelineEvent JSONL."""
        _require_native()
        mod = get_native_module()
        return mod.PolicyLab.to_timeline(events_jsonl)
